City-of-Helsinki/events-helsinki-monorepo
# LIIKUNTA-355, LIIKUNTA-357 | Add order by name/relevance/distance to unified search results

## Description

### feat(sports-helsinki): order unified search by name/relevance/distance
 - add geolocation provider and use it in app
 - add translations for geolocation related strings
 - add translations for order by parameters name/relevance/distance
 - use geolocation in unified search
 - add OrderBySelect component and add it to SearchResultsContainer
   - wrap the found events count and OrderBySelect in a
     styles.rowGroup div and that in a styles.row div, similar to
     liikunta-helsinki's styling in SearchList.tsx
   - add useHandleOrderChange hook
 - add an optional boolean argument showOrderBySelect to
   SearchResultsContainer's props and use it to conditionally show the
   OrderBySelect component
   - show the OrderBySelect component only in venue search's search page

components:
 - add GeolocationProvider
 - add GeolocationService
 - add useGeolocation
 - add Select + stylings
 - add SmallSpinner + stylings
 - add showToast

refactor:
 - move Coordinates type declaration to components
 - move Option type declaration to components

chore:
 - fix linting by removing unused EventTypeId imports from
   - AppConfig.ts
   - AppConfig.test.tsx

refs LIIKUNTA-355 (add unified search order by name/relevance/distance)
refs LIIKUNTA-357 (show notification if geolocation can not be used)

## Issues

### Closes

**[LIIKUNTA-355](https://helsinkisolutionoffice.atlassian.net/browse/LIIKUNTA-355)**
**[LIIKUNTA-357](https://helsinkisolutionoffice.atlassian.net/browse/LIIKUNTA-357)**

### Related

## Testing

### Automated tests

**Not much is done here**, **SonarCloud check currently fails** because of the code coverage of the new code is <80%. I'd rather merge this sooner than later and **add tests in another PR** if they are seen as important to be done before other functionality in this sprint.

### Manual testing

Tested that "padel" search gave the same results in production and in PR version when tested locally and the search results were sorted in the same order in both environments using name/relevance/distance choices (See screenshots below).

Also tried sorting the search results after denying Google Chrome the access to geolocation and got an error message (See screenshots below).

Checked that the order by component wasn't shown in the "Kurssihaku" (course search) for "Tapahtumat" (events) or "Harrastukset" (hobbies).

Checked that the translations for the geolocation permission denied error worked and that the translations for the order by choices worked (See screenshots below).

## Screenshots

Searched with "padel" term both in production version and PR version:

### Order by name in production

https://liikunta.hel.fi/search?q=padel&orderBy=name&orderDir=asc
![liikunta-hel-fi-order-by-name](https://user-images.githubusercontent.com/77663720/211569663-dbe6322d-b7ed-449e-a1d8-90db0d0148c8.png)

### Order by relevance in production

https://liikunta.hel.fi/search?q=padel&orderBy=relevance&orderDir=asc
![liikunta-hel-fi-order-by-relevance](https://user-images.githubusercontent.com/77663720/211569690-9c6312ee-c30f-4864-bd3e-622763eee843.png)

### Order by distance in production

https://liikunta.hel.fi/search?q=padel&orderBy=distance&orderDir=asc
![liikunta-hel-fi-order-by-distance](https://user-images.githubusercontent.com/77663720/211569709-dded50dd-58db-483d-aabe-03e6368600ae.png)

### Order by name in PR version

http://localhost:3000/fi/search?q=padel&orderBy=name&orderDir=asc
![pr-order-by-name](https://user-images.githubusercontent.com/77663720/211569726-0e8b0327-ee69-4ac6-afe4-cb7882e159fd.png)

### Order by relevance in PR version

http://localhost:3000/fi/search?q=padel&orderBy=relevance&orderDir=asc
![pr-order-by-relevance](https://user-images.githubusercontent.com/77663720/211569747-880fac85-29b1-4cb3-a3db-0a8137f2db86.png)

### Order by distance in PR version

http://localhost:3000/fi/search?q=padel&orderBy=distance&orderDir=asc
![pr-order-by-distance](https://user-images.githubusercontent.com/77663720/211569766-74f6171d-8607-409e-ba0e-c09f87aaae78.png)

### Error message if geolocation use was denied

When trying to order the search results according to distance and the user has denied the use of geolocation then an error is shown (Suomi/English/Svenska):
![geolocation-permission-denied-suomi](https://user-images.githubusercontent.com/77663720/211581157-6b17bd2a-5a2c-4d1c-96b4-3963272c979a.png)
![geolocation-permission-denied-english](https://user-images.githubusercontent.com/77663720/211581175-6203519d-4969-423b-8763-e061c91c24d9.png)
![geolocation-permission-denied-svenska](https://user-images.githubusercontent.com/77663720/211581188-3bb17c08-58de-47d6-9572-fe10cf97b146.png)

### Order by choices are translated (Suomi/English/Svenska)

![order-by-choices-suomi](https://user-images.githubusercontent.com/77663720/211580667-41efd5a1-2b4a-4d11-8f47-1c5d976ecc4c.png)
![order-by-choices-english](https://user-images.githubusercontent.com/77663720/211580684-5f8b98ce-3270-49d8-a98c-74f67fc708f1.png)
![order-by-choices-svenska](https://user-images.githubusercontent.com/77663720/211580700-bd22a213-7621-49e7-8e0b-a754ef5977f2.png)

## Additional notes


[LIIKUNTA-355]: https://helsinkisolutionoffice.atlassian.net/browse/LIIKUNTA-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
